### ISSUE TITLE
Sega CD - Chuck Rock game freezes in intro

### ISSUE THREAD

**Issue by KamiLordus** [ekeeke/Genesis-Plus-GX#560]
The game freezes at the beginning of the intro. You can hear the narrator and sounds, but the animation is stopped.

![Chuck Rock (USA)-240519-153357](https://github.com/ekeeke/Genesis-Plus-GX/assets/123302593/43722a5d-cfa2-4114-a391-e5deacec7fa4)

Windows 11 
Retroarch 1.18
Genesis Plus Gx core latest
Chuck Rock (USA) - chd file format

Retroarch log file:
[retroarch.log](https://github.com/ekeeke/Genesis-Plus-GX/files/15369588/retroarch.log)

**Comment by ekeeke** [comment]
I can confirm this is a regression introduced by commit https://github.com/ekeeke/Genesis-Plus-GX/commit/47761b9b8fd35eabd7b3ff3dc5ab0f54f1d95133

Most probably, this was working by chance in earlier versions but above commit slightly modified main-cpu execution timings just enough to break synchronization between main-cpu and sub-cpu. I will need to analyze this further.

Note: I also noticed that since this other commit (https://github.com/ekeeke/Genesis-Plus-GX/commit/dd619513a4a7136e7279cbf6adb660ac30598bdd), that game (and maybe others from Sony Imagesoft which share same boot routines) was taking longer time at the Sega logo so this is another thing to lookup.

**Comment by KamiLordus** [comment]
Wolfchild seems to have a very similar issue. The game freezes on the Sega logo, but if you wait a bit, or use the fast forward option in Retroarch, you can hear the music playing.

**Comment by jmaneyrol69** [comment]
It has actually become worse, even with the latest commit from https://github.com/libretro/Genesis-Plus-GX/commit/7b575132c6c7a97452d99d3222c6e006d987e710 : now all Core Design segaCD games I have tested so far have become unusable.

They all boot-loop right after the "Sega" boot logo. I have reproduced on Linux and Android with:

- Wolfchild (USA)
- Chuck Rock (USA)
- Wonderdog (USA)

My images are in CHD format.

Also, they work with other emus, such as PicoDrive.

**Comment by Freakservo** [comment]
Confirming Core bug -- Chuck 2 and BC Racers also freeze, and Thunderstrike _sometimes_ freezes after the opening cinematic (Other times at SEGA logo; About 50/50.)

**Comment by jmaneyrol69** [comment]
Some precision too ; the problem seems to differ depending on the system.

On Linux & Android, those games just loop boot right after the Sega loading logo (the one where we see Sonic).

On Windows, it goes further but then freezes, usually after the intro (Chuck Rock) or some time after the loading screen (WolfChild). Anyway, in all cases, Core games are unplayable on Sega CD for the moment with this core.

**Comment by ekeeke** [comment]
From experience, it differs depending on BIOS version used so most likely you have different BIOS configured on these different system.

Anyway, my spare time for working on this emulator is close to 0 these days and so far, there has been nobody else but me investigating emulation bugs so, unfortunately, it's not going to be solved soon. I looked briefly to it some times ago but it's kinda hard to figure out the root cause as this is likely a synchronization issue between the two CPUs, which is quite hard to debug.
 
For the time being, the only solution is to revert to an older version (previous to the commit indicated above).

**Comment by jmaneyrol69** [comment]
@ekeeke thank you for the details. I'm pretty sure I use the same BIOSes on all my systems. However, I have not tested other versions than US so far, I'll see what happens with EU/JP.

I hope you'll have enough time & support in the future to look further into the issue, and no worries, there are other options to run those games (though it will be nice when it is fixed ;-) ). Take care!

**Comment by Freakservo** [comment]
Okay then -- Workaround time! =)

After taking a bit to mess with the options, I discovered turning 'CD Access Time' off in the 'Emulation Hacks' menu gets these games moving.  I only ran them as far as their title-screens, but all Core-developed games seem to function with this one option deactivated.

**Comment by jmaneyrol69** [comment]
@Freakservo I was just going to report :wink: This actually fixes the boot loop on Linux & Android (I haven't tested on Windows yet, but you seem to be doing that yourself :slightly_smiling_face: ).

On those systems, I am having weird results with Wolfchild: with the US version, the game still crashes after the loading screen, right after the map. With the Japanese version, however, the game actually starts but display is completely corrupted, with weird blocks everywhere, missing tiles and sprites. Wonder Dog, on the other hand, seems to work well. I haven't tried any other Core games yet.

I have also found that Batman Returns (USA) is somehow affected by the problem: without the workaround, the game is stuck right after the SEGA logo. So this may even stretch to non-Core games as well :cry:

**Comment by Freakservo** [comment]
@jmaneyrol69 -
Can further-confirm with Wolfchild - The intro  can be skipped but the game freezes past the map and 'loading' screen.

Digging a little deeper, the EU-version of Chuck Rock does seem to work with the CD Access Time off -- Intro plays and I can load into the level.  I wonder how this affects other EU releases...  Including Batman and Wolfchild.

**Comment by Goratrix** [comment]
I just found this issue on my Anbernic handheld, tested a bunch of games thinking it may be a SD card issue first, here's the list of affected games I found so far:

AH3 - Thunderstrike (USA)
Batman Returns (USA)
BC Racers (USA)
Chuck Rock (USA)
Chuck Rock II - Son of Chuck (USA)
Fatal Fury Special (USA)
Garou Densetsu Special (Japan)
Hook (USA)
SoulStar (USA)

I also quickly tested these with 'CD Access Time' OFF, and they seem to launch fine into their intros, but I did not have time to actually play them further.

**Comment by jmaneyrol69** [comment]
@Goratrix turning CD Access Time OFF seems to workaround the issue on most of those games, but the result is variable depending on the version/BIOS, and the game (for instance, WolfChild won't work, no matter what).
Hopefully, ekeeke will have some time to look into the problem and release a fix...

**Comment by KamiLordus** [comment]
> I just found this issue on my Anbernic handheld, tested a bunch of games thinking it may be a SD card issue first, here's the list of affected games I found so far:
> 
> AH3 - Thunderstrike (USA) Batman Returns (USA) BC Racers (USA) Chuck Rock (USA) Chuck Rock II - Son of Chuck (USA) Fatal Fury Special (USA) Garou Densetsu Special (Japan) Hook (USA) SoulStar (USA)
> 
> I also quickly tested these with 'CD Access Time' OFF, and they seem to launch fine into their intros, but I did not have time to actually play them further.

Retroarch lets you save settings for each game separately. So disabling cd access time shouldn't be a problem here.

So most of the games you listed here work fine.

The games that clearly don't work are:
Chuck Rock (USA)
Wolf Child (USA)

From what I've seen @ekeeke doesn't have the time/will to work on this core, and he's also been struggling to get his github account back.

Too bad, because genesis plus gx is a very good core.

**Comment by sonninnos** [comment]
I did some investigating and noticed that doing this simple partial revert fixed most hangs with access time enabled:
```diff
diff --git a/core/cd_hw/cdd.c b/core/cd_hw/cdd.c
index 73ead1a1..07987348 100644
--- a/core/cd_hw/cdd.c
+++ b/core/cd_hw/cdd.c
@@ -1960,7 +1960,7 @@ void cdd_update(void)
   }

   /* check if seeking is pending */
-  if (cdd.pending)
+  if (scd.regs[0x38>>1].byte.h == CD_SEEK)
   {
     /* reset track index */
     int index = 0;
```

Chuck Rock at least still hangs though, so please do more testing if you can.

**Comment by sonninnos** [comment]
I decided to do this for the time being, since it seems to work well enough: https://github.com/libretro/Genesis-Plus-GX/pull/376

Won't fix Chuck Rock or Wolfchild hang, but do others that currently require disabling or enabling CD Access Time. So the option still works when it is actually required to be enabled, but it no longer needs to be disabled in some cases.

**Comment by ekeeke** [comment]
Hi, thanks for taking the time to look into this but unfortunately, this change will likely break sprites in Radical Rex into again (which is the initial reason these code changes about pending status were added in first place).

I need to find some time to investigate this regression with CD boot after this change when CD latency is enabled but haven't yet as it requires some deeper log analysis.

The hang during Chuck Rock intro has been caused by slight main cpu timing changes after introduction of 68k bus refresh delay emulation then later "improvment" of vint timing accuracy in h32 mode (reverting just these two commits makes the intro works again) so i suspect it's a main-cpu/sub-cpu synchronization issue that could be solved by adjusting hard-coded cycles values used to trigger cpu polling detection (which is also used to synchronize main and sub cpu execution) but again, i need more time to investigate as debugging sync issues is not easy.

**Comment by ekeeke** [comment]
1. After some debugging, i found the cause of the lock in Chuck Rock intro: as suspected, this is caused by some bugged code routine in original game that somehow ends up working (by luck) on real hardware. Below is the bugged code routine:

```
ROM:00002946 sub_2946:                               ; CODE XREF: sub_2946+8j
ROM:00002946                 move.w  $128(a5),d0
ROM:0000294A                 cmp.w   $128(a5),d0
ROM:0000294E                 beq.s   sub_2946
```

For those not familiar with 68K asm, this routine (executed in main-cpu RAM) waits for a value to be modified (by an interrupt routine) in RAM (pointed by a5)  but, instead of reading the RAM address only once then loops waiting for the value at this address to be changed (i.e branching to $294a instead of $2946), it continuously reads the address first then read it again immediately and waits for it to be changed between the two reads. This means this routine only works if the interrupt routine (that modifies the address) occurs exactly between the two reads, which eventually ends up occurring on real hardware because CPU execution timings slightly shift with time, while interrupts are always triggered at the exact same time. On emulators, timings are usually more deterministic so it's an example of a game that works by chance in emulators, as long as they trigger the interrupt at the right moment (the timings could be completely off in emulators but still get the game working as long it is triggered between those two reads). This is also an example of a game that get broken if you slightly modify CPU execution timings, even when trying to improve the accuracy (which was the case here when i added initial bus refresh delay emulation or later improved interrupt occurrence timing in H32 mode).

In commit https://github.com/ekeeke/Genesis-Plus-GX/commit/4529c8275b548385b4793375ef79016cb3c93609, I slightly modified bus refresh delay emulation so that CPU execution timings are less deterministic and it fixed the intro. It seems to have also fixed Wolfchild (although i did not analysed the original cause of the lock) and Pac-Man 2 arcade mini game (reported [here](https://github.com/libretro/Genesis-Plus-GX/issues/363).

2. Regarding the lock at Sega logo on boot (which also occurs in many other Sega CD games), it has been fixed by [this commit](https://github.com/ekeeke/Genesis-Plus-GX/commit/a65a0ffb538dd085a08526127c0cc4dd2eeb3155) which fixes a edge case with CD seek delay emulation, when a play/seek command is sent while another one is already in progress and under latency (which also explains why the issue was not occurring when CD latency emulation is disabled)

**Comment by sonninnos** [comment]
Excellent, thanks, I'll revert and fix the libretro repo.

**Comment by jmaneyrol69** [comment]
That's really great news! Thanks again for looking into it @ekeeke @sonninnos  ❤️
